### PR TITLE
recover resolver for app start

### PR DESCRIPTION
**Pull request by b-admike**
### Description

I had removed the code to handle the Promise resolution or rejection from `app.start` in #23, so adding it back.

### CODE REVIEW

File: src/apidefs/hello-world.api.ts [hunk 1]
```diff
@@ -69,7 +69,7 @@ export const helloWorldApi = {
             name: 'userInfo',
             in: 'body',
             description: 'The user model instance to create.',
-            required: false,
+            required: true,
```
Comment (kjdelisle) [inline src/apidefs/hello-world.api.ts]: Unrelated change

File: src/application.ts [hunk 2]
```diff
@@ -32,11 +32,18 @@ export class HelloWorldApp extends RepositoryMixin(Application) {
     server
       .bind(AuthenticationBindings.STRATEGY)
       .toProvider(MyAuthStrategyProvider);
-    await super.start();
-    let port = await server.get(RestBindings.PORT);
-    console.log(`HTTP server listening on port ${port}`);
-    console.log(
-      `Run \'curl localhost:${port}/helloworld?name=YOUR_NAME\' to try it out`,
+    await super.start().then(
```
Comment (kjdelisle) [inline src/application.ts]: This isn't necessary, you can just do this:
```ts
await super.start();
const port = await server.get(RestBindings.PORT);
console.log(`HTTP server listening on port ${port}`);
console.log(
  `Run \'curl localhost:${port}/helloworld?name=YOUR_NAME\' to try it out`,
);
```

If you want to handle errors from that, wrap it in try/catch.

File: src/application.ts [hunk 3]
```diff
@@ -32,11 +32,18 @@ export class HelloWorldApp extends RepositoryMixin(Application) {
...
+      async () => {
+        let port = await server.get(RestBindings.PORT);
+        console.log(`HTTP server listening on port ${port}`);
+        console.log(
+          `Run \'curl localhost:${port}/helloworld?name=YOUR_NAME\' to try it out`,
+        );
+      },
+      err => {
```
Comment (jannyHou) [inline src/application.ts]: A convention question: do we have any code convention about using 2nd param of `then()` or use `catch()` for error?

File: src/apidefs/hello-world.api.ts (hunk 1 above)
Comment (b-admike) [reply src/apidefs/hello-world.api.ts]: I was trying to sneak this one in, but alas! I'll create another PR for it :-)

File: src/application.ts (hunk 2 above)
Comment (b-admike) [reply src/application.ts]: Good point, but I thought the best practice in this case was to use the promise methods.

File: src/application.ts (hunk 3 above)
Comment (b-admike) [reply src/application.ts]: Good question. I should look through our code base to find a pattern!

File: src/application.ts (hunk 2 above)
Comment (kjdelisle) [reply src/application.ts]: You **are** using promise methods! Async and await are how you linearize the idea of asynchronous code. It's essentially a way of "blocking" execution of that function without having to implement callback hell/nested functions in `.then/.catch` blocks, and without needing to *actually* block the event loop.

You don't await functions that don't return Promises.

Comment (bajtos) [reply src/application.ts]: +1 to what Kevin said.

The only place where you need to explicitly add `.then` or `.catch` is the top-level script that's calling `start` from a plain (sync) method. When Node.js is loading a script from a file, it wraps the code in a module wrapper (see https://nodejs.org/api/modules.html#modules_the_module_wrapper):

```js
(function(exports, require, module, __filename, __dirname) {
// Module code actually lives in here
});
```

The wrapper function is sync, therefore you cannot use `await` and have to use `then`/`catch` instead.

This typically applies to the main file of your app - [src/index.ts](https://github.com/strongloop/loopback-next-hello-world/blob/175706c1bb6e0161080726400db477f0753dd40b/src/index.ts) in case of this repository:

```ts
if (require.main === module.parent) {
  // executed from the console
  const app = new HelloWorldApp();
  (async function main() {
    await app.start();
  })();
}
```

Here is an example how to modify that code to handle errors:

```ts
if (require.main === module.parent) {
  // executed from the console
  const app = new HelloWorldApp();
  app.start().catch(err => {
    console.error('Cannot start the application! ', err);
    process.exit(1);
  });
}
```

Comment (b-admike) [reply src/application.ts]: Thank you both. I have to catch the error at one or the other places right? I'm thinking catching it in `index.tx` would make more sense. Thoughts?

Comment (bajtos) [reply src/application.ts]: Yes, catch the error in one place only, and that's `index.ts`.

You don't want `app.start` to swallow errors (or convert them to console logs), because when you call `app.start` from a test and the start fails, then you want the test to fail too.

File: src/application.ts [hunk 4]
```diff
@@ -32,11 +32,16 @@ export class HelloWorldApp extends RepositoryMixin(Application) {
...
+      await super.start();
+      let port = await server.get(RestBindings.PORT);
+      console.log(`HTTP server listening on port ${port}`);
+      console.log(
+        `Run \'curl localhost:${port}/helloworld?name=YOUR_NAME\' to try it out`,
+      );
+    } catch (err) {
+      console.error('Cannot start the application! ', err);
+      process.exit(1);
+    }
```
Comment (bajtos) [inline src/application.ts]: Please revert changes in this file per my comment above.

You don't want `app.start()` to swallow errors (or convert them to a console log or a forceful process exit), because when you call `app.start()` from a test and the start fails, then you want the test to fail too, and let the test runner to report the test failure in a usual way.

File: src/index.ts [hunk 5]
```diff
@@ -10,7 +10,8 @@ import {RestServer, RestBindings} from '@loopback/rest';
 if (require.main === module.parent) {
   // executed from the console
   const app = new HelloWorldApp();
-  (async function main() {
-    await app.start();
-  })();
+  app.start().catch(err => {
+    console.error('Cannot start the application! ', err);
+    process.exit(1);
+  });
```
Comment (bajtos) [inline src/index.ts]: 👍

File: test/acceptance/error.handling.acceptance.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      app = new HelloWorldApp();
+      restServer = await app.getServer(RestServer);
+      await app.start();
+    } catch (err) {
+      console.log('err ', err);
+      console.log('keys ', Object.keys(err));
+      expect(err).to.match(/Cannot start the application!/);
+      expect(err.code).to.match(/EADDRINUSE/);
+    }
+  });
```
Comment (bajtos) [inline test/acceptance/error.handling.acceptance.ts]: Nice! 👍

File: package.json [hunk 7]
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "build": "tsc",
     "pretest": "npm run build",
-    "test": "mocha",
+    "test": "mocha -t 5000",
```
Comment (bajtos) [inline package.json]: Please move the timeout config to https://github.com/strongloop/loopback-next-hello-world/blob/master/test/mocha.opts

Do we really need 5s for every test? Can we increase the timeout only in the slow test?

```ts
it('Throws error for not starting', function () {
  this.timeout(5000);
  // etc.
}
```

(It's important to use a regular function, not an arrow function, so that `this` points to Mocha's context.)

**Comment by b-admike** [comment]
@bajtos @kjdelisle @jannyHou thanks for your feedback! PTAL.

File: test/acceptance/error-handling.acceptance.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import {expect} from '@loopback/testlab';
+import * as cp from 'child_process';
+import * as path from 'path';
+import * as http from 'http';
+
+describe('App error handler (acceptance)', () => {
+  let child: cp.SpawnSyncReturns<Buffer>;
```
Comment (bajtos) [inline test/acceptance/error-handling.acceptance.ts]: FWIW, since the `child` is used inside test only, I think there is no need to define it at `describe` suite level.

```ts
// inside the test
const child = cp.spawnSync(//...
```

File: test/acceptance/error-handling.acceptance.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  let httpServer: http.Server;
+
+  before(givenListeningServer);
+  after(stopHttpServer);
+
+  it('Throws error for not starting', function() {
```
Comment (bajtos) [inline test/acceptance/error-handling.acceptance.ts]: we usually start the test name with a lower case letter, `it('throws...`
